lib/concurrent/parallel.rb

```diff
@@ -0,0 +1,91 @@
+require 'concurrent/atomics'
+require 'concurrent/configuration'
+require 'concurrent/executors'
+
+module Concurrent
+
+  class Parallel < SimpleDelegator
+    module EnumerableInterface
+      module_function
+
+      def use(runner, methods)
+        enumerable_methods = Enumerable.instance_methods.map(&:to_s)
+        enumerable_methods << 'each'
+        methods.each do |method|
+          next unless enumerable_methods.include?(method)
+          class_eval <<-RUBY
+          def #{method}(*args, &block)
+            #{runner}(:#{method}, *args, &block)
+          end
+          RUBY
+        end
+      end
+
+      use :run_in_threads, %w(
+        all? any? count detect find find_index max_by min_by minmax_by none?
+        one? partition
+      )
+
+      use :run_in_threads_return_original, %w(
+        cycle each each_cons each_entry each_slice each_with_index enum_cons
+        enum_slice enum_with_index reverse_each zip
+      )
+
+      use :run_in_threads_return_parallel, %w(
```

Comment:
Yes, the idea was to have a fluent interface whereby one could parallize an enumerable and it would run all subsequent calls in parallel (where possible). The user could then get back the delegate at any time by calling `serial` on the parallel.

The advantage, as I saw it, isn't the saving of typing out the `parallel` method call. Rather, it gives an interesting story when it comes to integrating with existing code. For example: a `Repository`-like object could, rather than return a simple array of records, call `parallel` on that same array before handing it off. The result is that with the one declarative statement, large quantities of work across the application will be parallized (even across third-party code). This approach only works if `Parallel` quacks exactly like an `Enumerable` of course.

I completely agree this is perhaps several pegs higher on the astonishment scale than a explicit utility-like interface but I think it gains some fairly hefty advantages for the trade-off.